pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexType.java

```diff
@@ -0,0 +1,121 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.index;
+
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.creator.IndexCreationContext;
+import org.apache.pinot.segment.spi.index.column.ColumnIndexContainer;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.IndexConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+
+
+/**
+ * TODO: implement mutable indexes.
+ * @param <C> the class that represents how this object is configured.
+ * @param <IR> the {@link IndexReader} subclass that should be used to read indexes of this type.
+ * @param <IC> the {@link IndexCreator} subclass that should be used to create indexes of this type.
+ */
+public interface IndexType<C extends IndexConfig, IR extends IndexReader, IC extends IndexCreator> {
+
+  /**
+   * The unique id that identifies this index type.
+   * <p>The returned value for each index should be constant across different Pinot versions as it is used as:</p>
+   *
+   * <ul>
+   *   <li>They key used when the index is registered in IndexService.</li>
+   *   <li>The internal identification in v1 files and metadata persisted on disk.</li>
+   *   <li>The default toString implementation.</li>
+   *   <li>The key that identifies the index config in the indexes section inside
+   *   {@link org.apache.pinot.spi.config.table.FieldConfig}, although specific index types may choose to read other
+   *   names (for example, <code>inverted_index</code> may read <code>inverted</code> key.</li>
+   * </ul>
+   */
+  String getId();
+
+  Class<C> getIndexConfigClass();
+
+  /**
+   * The default config when it is not explicitly defined by the user.
+   */
+  C getDefaultConfig();
+
+  C getConfig(TableConfig tableConfig, Schema schema);
+
+  /**
+   * Optional method that can be implemented to ignore the index creation.
+   *
+   * Sometimes it doesn't make sense to create an index, even when the user explicitly asked for it. For example, an
+   * inverted index shouldn't be created when the column is sorted.
+   *
+   * Apache Pinot will call this method once all index configurations have been parsed and it is included in the
```

Comment:
`Apache Pinot` feels too generic to me here. Perhaps list some example callers of this method here, e.g. SegmentPreProcessor or SegmentCreator (iiuc) .